File: tilepuzzle.rb
class Direction
  Left = 0
  Right = 1
  Up = 2
  Down = 3
end

class TilePuzzle
    attr_reader :size, :tiles, :empty, :solution

    def initialize(n)
        @size = n
        @tiles =* (1..(n ** 2 -1))
        @tiles.push(0)
        @empty = n * n - 1
        @solution = Array.new(@tiles)
    end

    def shuffle()
        @tiles.shuffle!
        @empty = @tiles.index(0)
    end

    def solved?
        @tiles == @solution
    end

    def move(dirList)
        dirList.each {|dir| slide(dir)}
    end

    private 

    def slide(dir)
        newEmpty = empty

        case dir
        when Direction::Left
            newEmpty += 1
        when Direction::Right
            newEmpty -= 1
        when Direction::Up
            newEmpty -= size
        when Direction::Down
            newEmpty += @size
        end

        swapTiles(empty, newEmpty)
        @empty = newEmpty
    end

    def swapTiles(x, y)
        @tiles[x], @tiles[y] = @tiles[y], @tiles[x]
    end
end

t = TilePuzzle.new(2)
print(t.solved?)
t.move([:up])
print(t.solved?)

File: tilepuzzle.scala
sealed abstract class Direction
case object Left extends Direction
case object Right extends Direction
case object Up extends Direction
case object Down extends Direction

class TilePuzzle(n : Int)
{
    val size : Int = n
    private var empty : Int = size * size - 1
    private var tiles : Vector[Int] =
        (1 to (size * size - 1)).toVector :+ 0

    def shuffle ()
    {
        tiles = util.Random.shuffle(tiles)
        empty = tiles.indexOf(0)
    }

    def validMoves () : Vector[Direction] = 
    {
        var moves : Vector[Direction] = Vector()
        
        if (empty / size != 0)
        {
            moves = moves :+ Up
        }
        if (empty / size != size - 1)
        {
            moves = moves :+ Down
        }
        if (empty % size != 0)
        {
            moves = moves :+ Left
        }
        if (empty % size != size -1)
        {
            moves = moves :+ Right
        }

        moves
    }

    private def slide (dir : Direction)
    {
        var newEmpty : Int = empty

        dir match 
        {
            case Left => newEmpty -= 1
            case Right => newEmpty += 1
            case Down => newEmpty += size
            case Up => newEmpty -= size
        }

        swapTiles(empty, newEmpty)
        empty = newEmpty
    }

    private def swapTiles(x : Int, y : Int)
    {
        val xVal = tiles(x)
        val yVal = tiles(y)
        tiles = tiles.updated(x, yVal).updated(y, xVal)
    }
}